### PR TITLE
Feature/35869/st georges importer

### PR DESCRIPTION
**Pull request by lauramccluskey1**
What?
A new St Georges importer has been written to account for a new data format (see planio ticket 35869). The old St Georges importer is still needed, so it has been kept in the repository. 

Why?
St Georges changed the format of the data being sent. Therefore, a new importer had to be created to be able to parse this format.

How?
The importer was written based on the rules sent by Fiona.

Testing?
Unittests have been written. A QA of the variant counts has also been completed and signed off by Fiona.

### CODE REVIEW

File: lib/import/brca/scripts/bash/Import_all_brca_interactive.sh [hunk 1]
```diff
@@ -178,19 +178,32 @@ $BRAKE import:brca fname="$(echo "$x" | sed -e 's:.*pseudonymised_data/\(.*\):\1
...
 -not -path "*/2021/*" \
 -not -path "*/2022/*" \
 -not -path "*/2023/*" \
-! -name "c466c80823235315f4df98bb4a14c4937ee5cbc4_08.2020_STG HBOC PHE reported till 28082020.xlsx.pseudo")
+! -name "c466c80823235315f4df98bb4a14c4937ee5cbc4_08.2020_STG HBOC PHE reported till 28082020.xlsx.pseudo" \
+! -name "*HBOC*")
 do
 IFS="$OIFS"
-$BRAKE import:brca fname="$(echo "$x" | sed -e 's:.*pseudonymised_data/\(.*\):\1:')" prov_code=$PROV
+$BRAKE import:brca fname="$(echo "$x" | sed -e 's:.*pseudonymised_data/\(.*\):\1:')" prov_code='RJ7_2'
```
Comment (shilpigoeldev) [inline lib/import/brca/scripts/bash/Import_all_brca_interactive.sh:206]: Hard code prov code should be removed.

**Comment by lauramccluskey1** [comment]
> @lauramccluskey1 Handler tests are failing , can you check those please?

Thanks Shilpi! I have fixed the ones related to my tests, but its now giving integration errors. @bshand are you able to take a look at this?

File: lib/import/brca/providers/st_george/st_george_handler.rb [hunk 2]
```diff
@@ -1,436 +1,360 @@
...
-            if normal?(record)
-              normal_full_screen(genotype, genotypes)
-            elsif failed_test?(record)
-              failed_full_screen(genotype, genotypes)
-            elsif positive_cdna?(record) || positive_exonvariant?(record)
-              if record.raw_fields['genotype'].scan(CDNA_REGEX).size > 1
-                process_multiple_positive_variants(positive_genes, genotype, record, genotypes)
-              else
-                single_variant_full_screen(genotype, genotypes, positive_genes, record)
+            if genotype.attribute_map['genetictestscope'] == 'Targeted BRCA mutation test'
```
Comment (shilpigoeldev) [inline lib/import/brca/providers/st_george/st_george_handler.rb:33]: use base class methods -> genotype.targeted?

File: lib/import/brca/providers/st_george/st_george_handler.rb [hunk 3]
```diff
@@ -1,436 +1,360 @@
...
+              # determines the genes in the record and creates a genotype for each one
+              genes = process_genes_targeted(record)
+              # For each gene in the list of genes a new genotype will need to be created
+              genotypes = duplicate_genotype_targeted(genes, genotype)
+              genotypes.each do |single_genotype|
+                assign_test_status_targeted(single_genotype, record)
               end
+            elsif genotype.attribute_map['genetictestscope'] == 'Full screen BRCA1 and BRCA2'
```
Comment (shilpigoeldev) [inline lib/import/brca/providers/st_george/st_george_handler.rb:41]: genotype.full_screen?

File: lib/import/brca/providers/st_george/st_george_handler.rb [hunk 4]
```diff
@@ -1,436 +1,360 @@
...
-            if positive_genes.flatten.uniq.size > 1
-              variants = process_multi_genes_rec(record, positive_genes)
-            elsif positive_genes.flatten.uniq.size == 1
-              variants = process_uniq_gene_rec(record, positive_genes)
+          def duplicate_genotype_targeted(genes, genotype)
+            # When there is more than one gene listed a separate genotype needs to be created for each one
+            # The genotype is duplicated and the new gene is added to the duplicated genotype
+            # Each genotype is then added to the genoytypes list which this method then returns
+            genotypes = []
+            genes.each do |gene|
```
Comment (shilpigoeldev) [inline lib/import/brca/providers/st_george/st_george_handler.rb:101]: Not very sure with two loops logic here.. can methods like genes.flatten.compact_blank.uniq be used to reduce it to single genes array to loop?

**Review by shilpigoeldev — CHANGES_REQUESTED** [review]

File: lib/import/brca/providers/st_george/st_george_handler.rb (hunk 4 above)
Comment (lauramccluskey1) [reply lib/import/brca/providers/st_george/st_george_handler.rb]: see [b72975c](https://github.com/NHSDigital/data_management_system/pull/77/commits/b72975c9bb7ce4cd2ac1e0fa3bd23ff936a97093)

File: lib/import/brca/providers/st_george/st_george_handler.rb (hunk 3 above)
Comment (lauramccluskey1) [reply lib/import/brca/providers/st_george/st_george_handler.rb]: see [b72975c](https://github.com/NHSDigital/data_management_system/pull/77/commits/b72975c9bb7ce4cd2ac1e0fa3bd23ff936a97093)

File: lib/import/brca/providers/st_george/st_george_handler.rb (hunk 2 above)
Comment (lauramccluskey1) [reply lib/import/brca/providers/st_george/st_george_handler.rb]: updated in [b72975c](https://github.com/NHSDigital/data_management_system/pull/77/commits/b72975c9bb7ce4cd2ac1e0fa3bd23ff936a97093)

File: lib/import/brca/providers/st_george/st_george_handler.rb [hunk 5]
```diff
@@ -1,436 +1,360 @@
...
-            elsif deprecated_brca_genenames_moleculartestingtype?(record)
-              add_gene_from_deprecated_nomenclature_moleculartestingtype(genotype, record)
+            testscope = TEST_SCOPE_MAP[record.raw_fields['moleculartestingtype']]
+            if testscope == 'targeted'
+              genotype.add_test_scope(:targeted_mutation)
+            elsif testscope == 'fullscreen'
+              genotype.add_test_scope(:full_screen)
             else
-              @logger.debug "FAILED gene parse for: #{record.raw_fields['genotype']}"
+              genotype.add_test_scope(:no_genetictestscope)
```
[the remark this answers is not in the document]
Comment (lauramccluskey1) [reply lib/import/brca/providers/st_george/st_george_handler.rb]: updated in [b72975c](https://github.com/NHSDigital/data_management_system/pull/77/commits/b72975c9bb7ce4cd2ac1e0fa3bd23ff936a97093)

File: lib/import/brca/providers/st_george/st_george_handler.rb [hunk 6]
```diff
@@ -1,436 +1,360 @@
...
-            genotype.add_status(1)
-            genotypes.append(genotype)
-          end
+            return unless TEST_TYPE_MAP[record.raw_fields['moleculartestingtype']]
 
-          def process_failed_targeted(genotype, record, genotypes)
-            process_single_gene(genotype, record)
-            genotype.add_status(9)
-            genotypes.append(genotype)
+            genotype.add_molecular_testing_type(TEST_TYPE_MAP[record.raw_fields['moleculartestingtype']])
```
[the remark this answers is not in the document]
Comment (lauramccluskey1) [reply lib/import/brca/providers/st_george/st_george_handler.rb]: updated in [b72975c](https://github.com/NHSDigital/data_management_system/pull/77/commits/b72975c9bb7ce4cd2ac1e0fa3bd23ff936a97093)

File: lib/import/brca/scripts/bash/Import_all_brca_interactive.sh (hunk 1 above)
Comment (lauramccluskey1) [reply lib/import/brca/scripts/bash/Import_all_brca_interactive.sh]: see [8df17eb](https://github.com/NHSDigital/data_management_system/pull/77/commits/8df17ebc5b5bd20b2bc35fa463d0db292be166c6)

**Review by shilpigoeldev — CHANGES_REQUESTED** [review]

File: lib/import/brca/providers/st_george/st_george_handler.rb [hunk 7]
```diff
@@ -1,436 +1,354 @@
...
-            if record.raw_fields['genotype'].scan(CDNA_REGEX).size > 1
-              process_multiple_positive_variants(positive_genes, genotype, record, genotypes)
-            else
-              process_single_gene(genotype, record)
-              process_single_positive_variants(genotype, record)
-              process_single_protein(genotype, record)
-              genotypes.append(genotype)
-            end
-          end
+            genotype.add_test_scope(:no_genetictestscope)
```
Comment (shilpigoeldev) [inline lib/import/brca/providers/st_george/st_george_handler.rb:69]: Unknown scope has been added here @lauramccluskey1 but not considered while processing the records ? Should you be having an else condition for these cases to be processed in fill_genotypes?

File: lib/import/brca/providers/st_george/st_george_handler.rb [hunk 8]
```diff
@@ -1,436 +1,354 @@
...
-            add_variants_multiple_results(variants, genotype, genotypes) unless variants.nil?
-
+          def duplicate_genotype_targeted(genes, genotype)
+            # When there is more than one gene listed a separate genotype needs to be created for each one
+            # The genotype is duplicated and the new gene is added to the duplicated genotype
+            # Each genotype is then added to the genoytypes list which this method then returns
+            genotypes = []
+            genes.flatten.compact_blank.uniq.each do |gene_value|
+                # genotype only needs to be duplicated if there is more than one gene in the list
+                genotype = genotype.dup if genes.flatten.uniq.size > 1
```
[the remark this answers is not in the document]
Comment (lauramccluskey1) [reply lib/import/brca/providers/st_george/st_george_handler.rb:101]: I think we still need it as we only want genotype to be duplicated if there is more than one gene in the list. Unless you can see another way?

File: lib/import/brca/providers/st_george/st_george_handler.rb (hunk 7 above)
Comment (lauramccluskey1) [reply lib/import/brca/providers/st_george/st_george_handler.rb:69]: Great spot @shilpigoeldev! I have just spoke to Fiona about this as its not clear what to do in this instance for st georges. She said that as there were no cases with no genetictestscope, this should give a big error instead. This means that if we see any cases in future new rules can be created. So I will add this as a logger message. @NImeson this applies to colorectal too

File: lib/import/brca/providers/st_george/st_george_handler.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (lauramccluskey1) [reply lib/import/brca/providers/st_george/st_george_handler.rb]: Updated in [9104764](https://github.com/NHSDigital/data_management_system/pull/77/commits/9104764c8ae62c3c95b8e74cdc521840d4aabb83)

File: lib/import/brca/providers/st_george/st_george_handler.rb [hunk 9]
```diff
@@ -1,436 +1,354 @@
...
+            # Adds genotype to genotype list which is then outputted
+            genotypes = []
+            columns = ['gene', 'gene (other)', 'variant dna', 'test/panel']
+            counter = 0
+            columns.each do |column|
+              genes[column]&.each do |gene|
+                # don't need to duplicate genotype if only one gene
+                genotype = genotype.dup if counter.positive?
+                genotype.add_gene(gene)
+                # TODO: CHECK THIS!
```
[the remark this answers is not in the document]
Comment (lauramccluskey1) [reply lib/import/brca/providers/st_george/st_george_handler.rb]: Updated in [9104764](https://github.com/NHSDigital/data_management_system/pull/77/commits/9104764c8ae62c3c95b8e74cdc521840d4aabb83)

File: lib/import/brca/providers/st_george/st_george_handler.rb (hunk 7 above)
Comment (lauramccluskey1) [reply lib/import/brca/providers/st_george/st_george_handler.rb:67]: Logger error added in [9104764](https://github.com/NHSDigital/data_management_system/pull/77/commits/9104764c8ae62c3c95b8e74cdc521840d4aabb83)

File: lib/import/brca/providers/st_george/st_george_handler.rb [hunk 10]
```diff
@@ -1,436 +1,348 @@
...
-            elsif full_screen?(record)
-              genotype.add_test_scope(:full_screen)
-            elsif void_genetictestscope?(record)
-              @logger.debug 'Unknown moleculartestingtype'
-              genotype.add_test_scope(:no_genetictestscope)
+            genotypes.each do |single_genotype|
+              process_variants(single_genotype, record)
+              @persister.integrate_and_store(single_genotype)
             end
+            genotypes
```
Comment (shilpigoeldev) [inline lib/import/brca/providers/st_george/st_george_handler.rb:27]: Why would you need this @lauramccluskey1 ?

**Review by shilpigoeldev — CHANGES_REQUESTED** [review]

File: lib/import/brca/providers/st_george/st_george_handler.rb [hunk 11]
```diff
@@ -1,436 +1,348 @@
...
-            process_genetictestcope(genotype, record)
-            res = process_variants_from_record(genotype, record)
-            res.each { |cur_genotype| @persister.integrate_and_store(cur_genotype) }
-          end
-
-          def add_organisationcode_testresult(genotype)
-            genotype.attribute_map['organisationcode_testresult'] = '697N0'
-          end
+            # records using new importer should only have SRIs starting with V
+            return unless record.raw_fields['servicereportidentifier'].start_with?('V')
```
Comment (shilpigoeldev) [inline lib/import/brca/providers/st_george/st_george_handler.rb:15]: return statement should be the first one, no need to create genotype object if we are not processing file @lauramccluskey1

File: lib/import/brca/providers/st_george/st_george_handler.rb [hunk 12]
```diff
@@ -1,436 +1,348 @@
...
-                process_deprecated_gene(deprec_gene, positive_gene_rec) if deprec_gene.present?
+          def interrogate_variant_dna_column(record, genotype, genes, column, gene)
+            # For full screen tests only- add test status when variant dna column is not empty
+            if record.raw_fields['variant dna'].match(/Fail/ix)
+              genotype.add_status(9)
+            elsif record.raw_fields['variant dna'] == 'N'
+              genotype.add_status(1)
+            # variant dna is not '*Fail*', 'N' or null AND raw:gene is not null AND raw:gene (other) is null
+            # 2 (abnormal) for gene in raw:gene. 1 (normal) for all other genes.
+            elsif record.raw_fields['gene'].present? \
```
Comment (shilpigoeldev) [inline lib/import/brca/providers/st_george/st_george_handler.rb:231]: You don't need backslashes for continuation here (and throughout the file) , it can be written as -

record.raw_fields['gene'].present? &&
record.raw_fields['gene (other)'].blank?

**Review by shilpigoeldev — CHANGES_REQUESTED** [review]

File: lib/import/brca/providers/st_george/st_george_handler.rb (hunk 12 above)
Comment (lauramccluskey1) [reply lib/import/brca/providers/st_george/st_george_handler.rb:231]: If I remove the backslashes my tests fail and I get "syntax error, unexpected &&"

File: lib/import/brca/providers/st_george/st_george_handler.rb [hunk 13]
```diff
@@ -1,436 +1,348 @@
...
 
-          def positive_cdna?(record)
-            variant = record.raw_fields['genotype']
-            variant.scan(CDNA_REGEX).size.positive?
-          end
-
-          def positive_exonvariant?(record)
-            variant = record.raw_fields['genotype']
-            variant.scan(EXON_VARIANT_REGEX).size.positive?
+          def match_fail(gene, record, genotype)
```
[the remark this answers is not in the document]
Comment (lauramccluskey1) [reply lib/import/brca/providers/st_george/st_george_handler.rb:264]: Updated in [08d5202](https://github.com/NHSDigital/data_management_system/pull/77/commits/08d5202924c702b8cacba6321dc38912a1147ecf)

File: lib/import/brca/providers/st_george/st_george_handler.rb (hunk 11 above)
Comment (lauramccluskey1) [reply lib/import/brca/providers/st_george/st_george_handler.rb]: Updated in [08d5202](https://github.com/NHSDigital/data_management_system/pull/77/commits/08d5202924c702b8cacba6321dc38912a1147ecf)

File: lib/import/brca/providers/st_george/st_george_handler.rb (hunk 10 above)
Comment (lauramccluskey1) [reply lib/import/brca/providers/st_george/st_george_handler.rb:28]: I needed it to be able to test the unittests for that function

File: lib/import/brca/providers/st_george/st_george_handler.rb (hunk 12 above)
Comment (shilpigoeldev) [reply lib/import/brca/providers/st_george/st_george_handler.rb:232]: @lauramccluskey1 Did you move `&&` operator to first line ?, below will work

```
def interrogate_variant_dna_column(record, genotype, genes, column, gene)
            # For full screen tests only- add test status when variant dna column is not empty
            if record.raw_fields['variant dna'].match(/Fail/ix)
              genotype.add_status(9)
            elsif record.raw_fields['variant dna'] == 'N'
              genotype.add_status(1)
            # variant dna is not '*Fail*', 'N' or null AND raw:gene is not null AND raw:gene (other) is null
            # 2 (abnormal) for gene in raw:gene. 1 (normal) for all other genes.
            elsif record.raw_fields['gene'].present? && record.raw_fields['gene (other)'].blank?
              update_status(2, 1, column, 'gene', genotype)
            # variant dna is not '*Fail*', 'N' or null AND raw:gene is not null AND raw:gene (other) is not null
            # 2 (abnormal) for gene in raw:gene.
            # 9 (failed, genetic test) for any gene specified WITH 'Fail' in raw:gene (other).
            # 1 (normal) for all other genes
            elsif record.raw_fields['gene'].present? && record.raw_fields['gene (other)'].present?
              if column == 'gene'
                genotype.add_status(2)
              elsif column == 'gene (other)'
                match_fail(gene, record, genotype)
              else
                genotype.add_status(1)
              end
            # variant dna not '*Fail*', 'N' or null AND raw:gene is null AND raw:gene(other) not a single gene
            # If gene is specified in raw:variant dna, assign 2 (abnormal) for that gene and 1 (normal) for
            # all other genes.
            # Else interrogate raw:gene (other).
            elsif record.raw_fields['gene'].blank? && (genes['gene (other)'].blank? ||
                  genes['gene (other)'].length > 1) && !genes['variant dna'].nil? && 
                  genes['variant dna'].length >= 1
              update_status(2, 1, column, 'variant dna', genotype)
            # variant dna is not '*Fail*', 'N' or null AND raw:gene is null AND raw:gene (other) specifies a single gene
            # 2 (abnormal) for gene in raw:gene (other). 1 (normal) for all other genes.
            elsif record.raw_fields['gene'].blank? && !genes['gene (other)'].nil? && genes['gene (other)'].length == 1
              update_status(2, 1, column, 'gene (other)', genotype)
            end
          end
```

Comment (lauramccluskey1) [reply lib/import/brca/providers/st_george/st_george_handler.rb]: Thanks @shilpigoeldev - updated in [960d753](https://github.com/NHSDigital/data_management_system/pull/77/commits/960d753ac2d56ab6510a5e6b45508ca55d85128d)

File: lib/import/brca/providers/st_george/st_george_handler.rb [hunk 14]
```diff
@@ -184,137 +184,131 @@ def process_r208(record)
...
+          def assign_status_on_gene_other(record, genes, genotype, genotypes)
+            binding.pry if record.raw_fields['servicereportidentifier'] == 'V001777'
+            gene_other = record.raw_fields['gene (other)']
+            remaining_genes = @all_genes
+            case gene_other
+            when /Fail/i
+              process_failed_gene_other(genes, genotype, genotypes, remaining_genes)
+            when /\?\z/i
+              process_status_genes(genotype, 4, genes['gene'], genotypes)
+            when /^c\.|^Ex.*Del\z|^Ex.*Dup\z|^Het\sDel|^Het\sDup/ix
```
Comment (NImeson) [inline lib/import/brca/providers/st_george/st_george_handler.rb:279]: As confirmed by Fiona, regex needs to account for e.g. 'Hetdel' entries as well, it's currently assigning this as test status 4 but needs to be 2.

Comment (shilpigoeldev) [reply lib/import/brca/providers/st_george/st_george_handler.rb]: Thanks @NImeson , Updated and test added in [add51ac](https://github.com/NHSDigital/data_management_system/pull/77/commits/add51ac1fd9d6771e9538f3d2a9490eb2e3e314b)

**Review by NImeson — APPROVED** [review]
All looks good :-)
